testnet/stacks-node/src/tests/signer/commands/block_wait.rs

```diff
@@ -0,0 +1,122 @@
+use super::context::{SignerTestContext, SignerTestState};
+use crate::tests::signer::v0::{wait_for_block_pushed_by_miner_key, MultipleMinerTest};
+use madhouse::{Command, CommandWrapper};
+use proptest::prelude::{Just, Strategy};
+use std::sync::{atomic::Ordering, Arc, Mutex};
+
+pub struct WaitForBlockFromMiner1Command {
+    miners: Arc<Mutex<MultipleMinerTest>>,
+}
+
+impl WaitForBlockFromMiner1Command {
+    pub fn new(miners: Arc<Mutex<MultipleMinerTest>>) -> Self {
+        Self { miners }
+    }
+}
+
+impl Command<SignerTestState, SignerTestContext> for WaitForBlockFromMiner1Command {
+    fn check(&self, state: &SignerTestState) -> bool {
+        println!(
+            "Checking: Waiting for Nakamoto block from miner 1. Result: {:?}",
+            !state.mining_stalled
+        );
+        !state.mining_stalled
+    }
+
+    fn apply(&self, _state: &mut SignerTestState) {
+        println!("Applying: Waiting for Nakamoto block from miner 1");
+
+        let miners_arc = self.miners.clone();
+
+        let (miner_pk_1, last_confirmed_nakamoto_height_counter) = {
+            let miners = miners_arc.lock().unwrap();
+            let (miner_pk_1, _) = miners.get_miner_public_keys();
+            let last_confirmed_nakamoto_height = miners.get_primary_last_stacks_tip_counter();
+            (miner_pk_1, last_confirmed_nakamoto_height)
+        };
+
+        let last_confirmed_height = last_confirmed_nakamoto_height_counter
+            .0
+            .load(Ordering::SeqCst);
+        let expected_height = last_confirmed_height + 1;
+
+        println!(
+            "Waiting for Nakamoto block {} pushed by miner 1",
+            expected_height
+        );
+
+        let _miner_1_block = wait_for_block_pushed_by_miner_key(30, expected_height, &miner_pk_1)
+            .expect(&format!("Failed to get block {}", expected_height));
+    }
+
+    fn label(&self) -> String {
+        "WAIT_FOR_BLOCK_FROM_MINER_1".to_string()
+    }
```

Comment:
Is there a way we can add something like a `derive` macro that eliminates the need to write this? This is relatively verbose for something that is just the upper-snake-case version of the struct's name.

Reply:
> Dynamic fields are needed, because in a long scenario this can make any debugging process/shrinking easier.

Yeah that makes sense, but then IMO there should be a default implementation that just returns the struct name, and then you can optionally define `label(&self) -> String`. But I also agree that this is nit-picky and we can easily refine this in the future with macros.